autopush/tests/test_integration.py

```diff
@@ -0,0 +1,629 @@
+            if use_header:
+                headers = {"Content-Type": "application/x-www-form-urlencoded"}
+            else:
+                headers = {}
+            method = "PUT"
+
+        log.debug("%s body: %s", method, body)
+        http.request(method, url.path, body, headers)
+        resp = http.getresponse()
+        log.debug("%s Response: %s", method, resp.read())
+        eq_(resp.status, status)
+        if self.use_webpush and ttl != 0:
+            assert(resp.getheader("Location", None) is not None)
+
+        # Pull the notification if connected
+        if self.ws and self.ws.connected:
+            result = json.loads(self.ws.recv())
+            return result
+
+    def get_notification(self, timeout=0.2):
+        self.ws.settimeout(timeout)
+        try:
+            d = self.ws.recv()
+            log.debug("Recv: %s", d)
+            return json.loads(d)
+        except:
+            return None
+
+    def ping(self):
+        log.debug("Send: %s", "{}")
+        self.ws.send("{}")
+        result = self.ws.recv()
+        log.debug("Recv: %s", result)
+        eq_(result, "{}")
+        return result
+
+    def ack(self, channel, version):
+        msg = json.dumps(dict(messageType="ack",
+                              updates=[dict(channelID=channel,
+                                            version=version)]))
+        log.debug("Send: %s", msg)
+        self.ws.send(msg)
+
+    def disconnect(self):
+        self.ws.close()
+
+
+class IntegrationBase(unittest.TestCase):
+    def setUp(self):
+        import cyclone.web
+        from autobahn.twisted.websocket import WebSocketServerFactory
+        from autopush.main import skip_request_logging
+        from autopush.endpoint import (EndpointHandler, RegistrationHandler)
+        from autopush.settings import AutopushSettings
+        from autopush.websocket import (
+            SimplePushServerProtocol,
+            RouterHandler,
+            NotificationHandler,
+        )
+        settings = AutopushSettings(
+            hostname="localhost",
+            statsd_host=None,
+            endpoint_port="9020",
+            router_port="9030"
+        )
+
+        # Websocket server
+        self._ws_url = "ws://localhost:9010/"
+        factory = WebSocketServerFactory(self._ws_url)
+        factory.protocol = SimplePushServerProtocol
+        factory.protocol.ap_settings = settings
+        factory.setProtocolOptions(
+            webStatus=False,
+            maxFramePayloadSize=2048,
+            maxMessagePayloadSize=2048,
+            openHandshakeTimeout=5,
+        )
+        settings.factory = factory
+        self.websocket = reactor.listenTCP(9010, factory)
+
+        # Websocket HTTP router
+        # Internal HTTP notification router
+        r = RouterHandler
+        r.ap_settings = settings
+        n = NotificationHandler
+        n.ap_settings = settings
+        ws_site = cyclone.web.Application([
```

Comment:
It'd be nice if these were unified with the definitions in `main.py`, so we'd only need to change them in one place. OTOH, apart from adding message `DELETE` and receipts, we won't be changing them frequently.